corehq/apps/geospatial/reports.py

```diff
@@ -116,40 +115,148 @@ class CaseGroupingReport(BaseCaseMapReport):
+        polygon (or other GeoShape) is passed in the "feature" GET
+        param, adds that as a filter to ``query``.
+        """
+        # NOTE: Expects GeoShape in request.GET['feature']
+        if self.request.GET.get('feature'):
+            case_property = get_geo_case_property(self.domain)
+            geojson = json.loads(self.request.GET['feature'])
             shape = geojson_to_es_geoshape(geojson)
             relation = 'within' if shape['type'] == 'polygon' else 'intersects'
-            query.nested(
-                CASE_PROPERTIES_PATH,
-                filters.geo_shape(
+            if additional_filters:
+                additional_filters.append(filters.geo_shape(
                     field=case_property,
                     shape=shape,
                     relation=relation,
+                ))
+            else:
+                query.nested(
+                    CASE_PROPERTIES_PATH,
+                    filters.geo_shape(
+                        field=case_property,
+                        shape=shape,
+                        relation=relation,
+                    )
                 )
+        if additional_filters:
+            query.nested(
+                CASE_PROPERTIES_PATH,
+                additional_filters
             )
+        return query
+
+    def _aggregate_query(self, query):
+        """
+        Returns ``query`` with geohash grid aggregation applied.
 
-        # Apply geohash grid aggregation
-        if 'precision' in self.request.GET:
-            precision = self.request.GET['precision']
+        Also determines ``self._precision`` if it is not set.
+        """
+        case_property = get_geo_case_property(self.domain)
+
+        if self.request.GET.get('precision'):
+            self._precision = int(self.request.GET['precision'])
         else:
-            precision = find_precision(query, case_property)
+            # TODO: What is the best way to send this value to the
+            #       browser, for the browser to pass it back as a GET
+            #       param across page requests?
+            self._precision = find_precision(query, case_property)
 
-        query = apply_geohash_agg(query, case_property, precision)
-        return query
+        return apply_geohash_agg(query, case_property, self._precision)
 
     def _get_geo_point(self, case):
         case_obj = wrap_case_search_hit(case)
         geo_case_property = get_geo_case_property(case_obj.domain)
         geo_point = case_obj.case_json.get(geo_case_property)
         return geo_point
 
+    @property
+    def rows(self):
+        """
+        Returns paginated cases
+        """
+        # Pagination uses a three-step process:
+        # 1. Cases are filtered by `self._build_query()`
+        # 2. The filtered cases are aggregated into buckets by
+        #    `self._aggregate_query()`.
+        # 3. The cases of the selected page may span buckets. The page's
```

Comment:
Is this needed? I believe we are to simply return cases of a single bucket or geohash.

Reply:
Yeah, this is the crucial part.

Imagine three buckets, each containing 33 cases. The user says they want to see 50 cases per page. What do we do?

1. Ignore the user, and show them 33?
2. Respect the user, and show them 50?

The code goes for option 2. In order to do that, page 1 will have 33 cases from the first bucket, and 17 from the second bucket. Page 2 will have 16 cases from the second bucket and 33 from the third.

Reply:
To be clear, I'm welcoming correction. Should the code go for option 1 instead?

Reply:
I am confused to why we are doing this. I assume we would show all cases in a bucket/group at once and not across buckets and each page is simply an identifier for the bucket.
This is what the requirements says "Clicking on any geohash page should trigger a request to the backend and fetch the list of cases for that geohash" so we are not expected to return cases across buckets.

Reply:
> each page is simply an identifier for the bucket

At first that is the approach I took.

But a bucket can have 10,000 cases. That's a lot of rows! It felt like a very poor user experience to me, which is why I transitioned to something that felt more consistent with all the other reports in HQ.

:thinking: But maybe I've been thinking about this wrong. These are not rows in a table, are they? They are pins on a map. So 10,000 pins is fine ... right?

Let me do that. I'll also look into removing the entries per page select box.

Reply:
902230736104db28aa27126df7749e9faa560d4f